### PR TITLE
remove do_*() methods from Program class

### PR DESCRIPTION
**Pull request by agb94**

### CODE REVIEW

File: pyggi/line/line.py [hunk 1]
```diff
@@ -37,7 +23,9 @@ def __init__(self, target, ingredient):
         self.ingredient = ingredient
 
     def apply(self, program, new_contents, modification_points):
-        return program.do_replace(self, new_contents, modification_points)
+        engine = program.engines[self.target[0]]
+        assert engine == program.engines[self.ingredient[0]]
```
Comment (bloa) [inline pyggi/line/line.py]: Unless the ingredient is modified after the edit creation (and even so), the assert can be moved in the `create()` method so it is checked only once.

File: pyggi/line/line.py [hunk 2]
```diff
@@ -9,20 +9,6 @@ class LineProgram(AbstractProgram):
     def get_engine(cls, file_name):
```
Comment (bloa) [inline pyggi/line/line.py]: Both LineProgram and TreeProgram have only a single method, which is rather pointless.
I think this can now be moved directly in the main program entry point (e.g., improve.py).

File: pyggi/line/line.py [hunk 3]
```diff
@@ -37,7 +23,9 @@ def __init__(self, target, ingredient):
         self.ingredient = ingredient
 
     def apply(self, program, new_contents, modification_points):
-        return program.do_replace(self, new_contents, modification_points)
+        engine = program.engines[self.target[0]]
+        assert engine == program.engines[self.ingredient[0]]
+        return engine.do_replace(program, self, new_contents, modification_points)
```
Comment (bloa) [inline pyggi/line/line.py]: Where is `do_replace()` defined?
It lacks an abstract class between the granularity-agnostic AbstractEngine and its implementation that would declare the methods `do_replace()` etc. This is even more visible for the tree-level granularity.

File: pyggi/line/line.py (hunk 2 above)
Comment (agb94) [reply pyggi/line/line.py]: `get_engine` defines the semantic of each class `LineProgram` and `TreeProgram`. I think these class-engine mappings should be defined inside of PyGGi, not the user-side program.

File: pyggi/line/line.py (hunk 3 above)
Comment (agb94) [reply pyggi/line/line.py]: Ok, then we should add one more class, for example, `AbstractTreeEngine`, between the `AbstractEngine` and `AstorEngine`/`SrcMLEngine`.

File: pyggi/line/line.py (hunk 1 above)
Comment (agb94) [reply pyggi/line/line.py]: but the edit instance can be created via not only `create` but also `__init__` method. how about add `is_valid_for(self, program)` methods to the edit classes and check the validity in the method `add` of `Patch`?

Comment (bloa) [reply pyggi/line/line.py]: Unless mistaken as for now the only way PyGGI instantiate new edits is through the `create` method, but you are right the assert is probably better moved directly in `__init__`.
In my opinion checking validity in the `add` method of `Patch` would only make sense if it depended of the other edits.

File: pyggi/line/line.py (hunk 2 above)
Comment (bloa) [reply pyggi/line/line.py]: In any case the user still has to chose which engine he wants to use.
Providing the classes `LineProgram` and `TreeProgram` solely containing the `get_engine` method is simply akin is replace an explicit decision (which engine to use) by another, blurrier, explicit decision (which type of program to use).

My last argument is that it is much cleaner to conditionally change a method content that the inheritance of a class, for example for an entry point allowing the user to use a CLI options to use either line or tree granularity level (albeit I see that Python allows some heretical constructs such as `class MyProgram(LineProgram if args.line else TreeProgram)`).

File: pyggi/line/line.py (hunk 3 above)
Comment (bloa) [reply pyggi/line/line.py]: Yes, we need "AbstractEngine > (Abstract)TreeEngine > AstorEngine" for the exact same reasons we had "AbstractProgram > (Abstract)TreeProgram > AstorProgram".

File: pyggi/line/line.py (hunk 2 above)
Comment (bloa) [reply pyggi/line/line.py]: Furthermore, I am thinking that removing or renaming tags in the XML tree in order to have a shorter list of modification points will necessarily require sub-classing the XmlEngine to overwrite its `get_content` method (or monkeypatch it, wich is not desirable).
In this case, it is utterly pointless to use TreeProgram because its `get_engine` method will be ignored (to use say MyXmlEngine rather than XmlEngine).

File: pyggi/line/line.py (hunk 1 above)
Comment (agb94) [reply pyggi/line/line.py]: @bloa But there's one problem😞. We cannot check the validity in `__init__` since we don't know the program instance to which the edit will be applied at that moment.

File: pyggi/line/line.py (hunk 3 above)
Comment (agb94) [reply pyggi/line/line.py]: Ok. Will fix it soon.

Comment (agb94) [reply pyggi/line/line.py]: commit: https://github.com/coinse/pyggi/pull/4/commits/8be595dd22973464e0426d7748aa395ae616057a

File: pyggi/line/line.py (hunk 2 above)
Comment (agb94) [reply pyggi/line/line.py]: I don't think it is desirable to define another Engine inheriting `SrcMLEngine`. We can rather create another (lower) level of subclasses (XmlNode(Tag)Class or something) and connect them in the main engine class as it would be more clear that all the SrcML handling parts are executed via the `SrcMLEngine`. 
Then, we just implement the code manipulators (java, c++, ,.. tree with `SrcMLEngine` and python code with `AstorEngine`) and tie them together to ultimately provide the `TreeProgram` and then users don't even need to care about what's behind `TreeProgram` or what `SrcML` or `Astor` is. So what users can/need to see are only `TreeProgram`(that supports `python`, `java`, and etc.) and the available edit operators.

Comment (agb94) [reply pyggi/line/line.py]: Or, do you mean providing two versions of SrcmlEngine?

File: pyggi/line/line.py (hunk 1 above)
Comment (bloa) [reply pyggi/line/line.py]: You are right again, the program is not known in `__init__`, this is quite unfortunate.
The different possibilities are as follows:
- leave the assert in the `apply` method, which means coherence is checked again and again for no particular reason;
- move the assert in `create`, meaning that edits directly user-instantiated (via `__init__`) are not validated
- adding the program as optional argument to `__init__`, for the only sake of validation
- have a dedicated validation method, called from somewhere else (as you mentioned, presumably `add` in Patch)
- simply remove the assert

Ideally, I would rather have a system that can be proved to not fail than one that checks that it did indeed not fail.
For an edit to be valid, we need first existence of the modification points, validity of the operation for the associated engines, and coherence of the engines if multiple modification points are involved.
Existence can be more or less inferred from the `random_target` method, i don't think validity is neither checked nor inferred, and coherence is the present assert.
(Actually, on top of that, validity inside a patch also depends on the previous edits.)

The best solution I can think of would be to restrict in the `create` method (or in the `random_target` nethod really) the target file to one for which the engine is compatible (and we now have a shared abstract engine class, perfect).
Basically, in simplified pseudo code:
```
def create(cls, program):
    target = program.random_target(AbstractTreeEngine)
    ingredient = program.random_target(program.get_engine(target))
    return cls(target, ingredient)
```
This would ensure that PyGGI will not fail due to incoherent engines.
As for edits directly created using `__init__`, I would simply do nothing: if a user wants to bypass `create` then it is its own responsibility not to provide invalid arguments (in which case PyGGI will fail anyway, assert or not).

File: pyggi/line/line.py (hunk 2 above)
Comment (bloa) [reply pyggi/line/line.py]: An engine is a collection of methods describing how to deal with a type of file (and I still think everything should be class methods). In particular, it hides the method that parses the original code and output its representation.
For xml, the representation is a tree with (far too) many tags, and I suppose it will be absolutely necessary to remove or rename some of the tags.
One solution would be to use class variables to setup the class so that this is done automatically during parsing; this is _horrible_.
Another much simpler and cleaner solution would be to subclass the xml engine to provide a new parse method (basically, a `super` and then the modification of the tree), and provide this engine to PyGGI instead.

File: pyggi/line/line.py (hunk 1 above)
Comment (agb94) [reply pyggi/line/line.py]: please check `https://github.com/coinse/pyggi/pull/4/commits/a37dcb55d7616e8f810408d5137ca85067f02493`.

File: pyggi/line/line.py (hunk 2 above)
Comment (agb94) [reply pyggi/line/line.py]: > An engine is a collection of methods describing how to deal with a type of file (**and I still think everything should be class methods**). In particular, it hides the method that parses the original code and output its representation.

Ok, I'll change it.

And why don't we provide only the XML engine with the new parse method? I mean, we can just convert the heavy XML representations into the simpler version of representation in that engine without requiring any subclasses. Or maybe you can just make commits, and then we'll see.
